File: ocaml/planet-time/lib/orbital.ml
(* orbital.ml — Orbital mechanics for the Interplanetary Time Library
   Story 18.18 — OCaml port of planet-time
   Exact port of planet-time.js / libinterplanet.c orbital functions. *)

let pi     = 4.0 *. atan 1.0
let two_pi = 2.0 *. pi
let d2r    = pi /. 180.0

(* ── Leap-second / TT helpers ────────────────────────────────────────────── *)

(* TAI − UTC in seconds for a given UTC millisecond timestamp *)
let tai_minus_utc (utc_ms : float) : int =
  let offset = ref 10 in
  Array.iter (fun (s, t_ms) ->
    if utc_ms >= t_ms then offset := s
  ) Constants.leap_secs;
  !offset

(* Convert UTC milliseconds to Terrestrial Time Julian Ephemeris Day.
   TT = UTC + (TAI−UTC) + 32.184 s *)
let jde (utc_ms : float) : float =
  let tt_ms = utc_ms +. (float_of_int (tai_minus_utc utc_ms) +. 32.184) *. 1000.0 in
  Constants.unix_epoch_jd +. tt_ms /. 86400000.0

(* Julian centuries from J2000.0 *)
let jc (utc_ms : float) : float =
  (jde utc_ms -. Constants.j2000_jd) /. 36525.0

(* ── Julian Day (calendar-based) ─────────────────────────────────────────── *)

(* Compute Julian Day from calendar date and time.
   Uses the standard algorithm (integer arithmetic for Gregorian calendar). *)
let julian_day ~year ~month ~day ~hour ~minute ~(second : float) : float =
  let y = if month <= 2 then year - 1 else year in
  let m = if month <= 2 then month + 12 else month in
  let a = y / 100 in
  let b = 2 - a + (a / 4) in
  let jd_day = float_of_int (int_of_float (365.25 *. float_of_int (y + 4716)))
             +. float_of_int (int_of_float (30.6001 *. float_of_int (m + 1)))
             +. float_of_int day +. float_of_int b -. 1524.5 in
  let time_frac = (float_of_int hour +. float_of_int minute /. 60.0 +. second /. 3600.0)
                  /. 24.0 in
  jd_day +. time_frac

(* ── Kepler solver (Newton-Raphson, 50 iterations) ───────────────────────── *)

let solve_kepler (m_anom : float) (ecc : float) : float =
  let e = ref m_anom in
  for _ = 1 to 50 do
    let de = (m_anom -. !e +. ecc *. sin !e) /. (1.0 -. ecc *. cos !e) in
    e := !e +. de
  done;
  !e

(* ── Mean longitude ──────────────────────────────────────────────────────── *)

(* Mean longitude of a body in degrees for a given Julian Day.
   Body index: 0=Mercury..7=Neptune (8=Moon uses Earth's elements). *)
let mean_longitude ~(body : int) ~(jd : float) : float =
  let idx = if body = 8 then 2 else body in
  let el = Constants.orb_elems.(idx) in
  let t = (jd -. Constants.j2000_jd) /. 36525.0 in
  let l_deg = el.Constants.l0 +. el.Constants.dl *. t in
  (* normalise to [0, 360) *)
  let l = l_deg -. 360.0 *. floor (l_deg /. 360.0) in
  l

(* ── True anomaly ────────────────────────────────────────────────────────── *)

(* Solve for true anomaly given mean anomaly (radians) and eccentricity.
   Uses Newton-Raphson via Kepler's equation.  Returns radians. *)
let true_anomaly ~(mean_anomaly : float) ~(eccentricity : float) : float =
  let e_anom = solve_kepler mean_anomaly eccentricity in
  2.0 *. atan2
    (sqrt (1.0 +. eccentricity) *. sin (e_anom /. 2.0))
    (sqrt (1.0 -. eccentricity) *. cos (e_anom /. 2.0))

(* ── Ecliptic longitude ──────────────────────────────────────────────────── *)

(* Ecliptic (heliocentric) longitude of a body in degrees. *)
let ecliptic_longitude ~(body : int) ~(jd : float) : float =
  let idx = if body = 8 then 2 else body in
  let el  = Constants.orb_elems.(idx) in
  let t   = (jd -. Constants.j2000_jd) /. 36525.0 in
  let l   = ((el.Constants.l0 +. el.Constants.dl *. t) *. d2r) in
  let l   = (mod_float l two_pi +. two_pi) |> fun x -> mod_float x two_pi in
  let om  = el.Constants.om0 *. d2r in
  let m   = (mod_float (l -. om) two_pi +. two_pi) |> fun x -> mod_float x two_pi in
  let e_anom = solve_kepler m el.Constants.e0 in
  let v   = 2.0 *. atan2
              (sqrt (1.0 +. el.Constants.e0) *. sin (e_anom /. 2.0))
              (sqrt (1.0 -. el.Constants.e0) *. cos (e_anom /. 2.0)) in
  let lon = (mod_float (v +. om) two_pi +. two_pi) |> fun x -> mod_float x two_pi in
  lon *. 180.0 /. pi

(* ── Heliocentric position ───────────────────────────────────────────────── *)

(* Heliocentric (x, y, z) position in AU.
   z is always 0.0 (ecliptic plane computation).
   Returns (x, y, z). *)
let heliocentric_pos ~(body : int) ~(jd : float) : float * float * float =
  let idx = if body = 8 then 2 else body in
  let el  = Constants.orb_elems.(idx) in
  let t   = (jd -. Constants.j2000_jd) /. 36525.0 in
  let l   = ((el.Constants.l0 +. el.Constants.dl *. t) *. d2r) in
  let l   = (mod_float l two_pi +. two_pi) |> fun x -> mod_float x two_pi in
  let om  = el.Constants.om0 *. d2r in
  let m   = (mod_float (l -. om) two_pi +. two_pi) |> fun x -> mod_float x two_pi in
  let e_anom = solve_kepler m el.Constants.e0 in
  let v   = 2.0 *. atan2
              (sqrt (1.0 +. el.Constants.e0) *. sin (e_anom /. 2.0))
              (sqrt (1.0 -. el.Constants.e0) *. cos (e_anom /. 2.0)) in
  let r   = el.Constants.a *. (1.0 -. el.Constants.e0 *. cos e_anom) in
  let lon = (mod_float (v +. om) two_pi +. two_pi) |> fun x -> mod_float x two_pi in
  let x   = r *. cos lon in
  let y   = r *. sin lon in
  (x, y, 0.0)

(* Heliocentric distance in AU for a body at a UTC ms timestamp *)
let helio_r_au ~(body : int) ~(utc_ms : float) : float =
  let idx = if body = 8 then 2 else body in
  let el  = Constants.orb_elems.(idx) in
  let t   = jc utc_ms in
  let l   = ((el.Constants.l0 +. el.Constants.dl *. t) *. d2r) in
  let l   = (mod_float l two_pi +. two_pi) |> fun x -> mod_float x two_pi in
  let om  = el.Constants.om0 *. d2r in
  let m   = (mod_float (l -. om) two_pi +. two_pi) |> fun x -> mod_float x two_pi in
  let e_anom = solve_kepler m el.Constants.e0 in
  el.Constants.a *. (1.0 -. el.Constants.e0 *. cos e_anom)

(* ── Light travel time ───────────────────────────────────────────────────── *)

(* One-way light travel time in seconds between two bodies at a given UTC ms *)
let light_travel_time ~(body1 : int) ~(body2 : int) ~(jd : float) : float =
  let (x1, y1, _) = heliocentric_pos ~body:body1 ~jd in
  let (x2, y2, _) = heliocentric_pos ~body:body2 ~jd in
  let dx = x1 -. x2 in
  let dy = y1 -. y2 in
  let dist_au = sqrt (dx *. dx +. dy *. dy) in
  dist_au *. Constants.au_seconds

(* Light travel time using utc_ms rather than JD (for fixture testing) *)
let light_travel_time_ms ~(body1 : int) ~(body2 : int) ~(utc_ms : float) : float =
  let jd = jde utc_ms in
  light_travel_time ~body1 ~body2 ~jd

File: rust/planet-time/src/orbital.rs
//! Orbital mechanics — heliocentric positions, light-travel time, line-of-sight.
//! Mirrors the orbital section of planet-time.js.

use crate::constants::{
    AU_SECONDS, DEG, J2000_MS, LEAP_SECS, ORB_ELEMS, Planet,
};
use std::f64::consts::PI;

// ── TAI − UTC ──────────────────────────────────────────────────────────────────

/// Returns TAI − UTC in seconds for the given UTC millisecond timestamp.
pub fn tai_minus_utc(utc_ms: i64) -> f64 {
    let mut ls = 10i64;
    for &(ts, s) in LEAP_SECS.iter() {
        if utc_ms >= ts { ls = s; }
    }
    ls as f64
}

// ── Julian date helpers ───────────────────────────────────────────────────────

/// Julian Ephemeris Day (TT) from UTC milliseconds.
pub fn jde(utc_ms: i64) -> f64 {
    let tai = utc_ms as f64 + tai_minus_utc(utc_ms) * 1000.0;
    let tt  = tai + 32_184.0;  // TT = TAI + 32.184 s (in ms)
    2_440_587.5 + tt / 86_400_000.0
}

/// Julian centuries from J2000.0 (TT).
pub fn jc(utc_ms: i64) -> f64 {
    (jde(utc_ms) - 2_451_545.0) / 36_525.0
}

// ── Kepler equation ───────────────────────────────────────────────────────────

/// Solve Kepler's equation M = E − e·sin(E) for eccentric anomaly E (radians).
/// Newton-Raphson, tolerance 1 × 10⁻¹².
pub fn kepler_e(m: f64, e: f64) -> f64 {
    let mut big_e = m;
    loop {
        let delta = (m - big_e + e * big_e.sin()) / (1.0 - e * big_e.cos());
        big_e += delta;
        if delta.abs() < 1e-12 { break; }
    }
    big_e
}

// ── Heliocentric position ─────────────────────────────────────────────────────

/// Heliocentric ecliptic position (AU, radians).
#[derive(Debug, Clone, Copy)]
pub struct HelioPos {
    pub x: f64,
    pub y: f64,
    pub r: f64,   // distance from Sun in AU
    pub lon: f64, // ecliptic longitude (radians, 0–2π)
}

/// Compute heliocentric position for `planet` at `utc_ms`.
pub fn helio_pos(planet: Planet, utc_ms: i64) -> HelioPos {
    let t  = jc(utc_ms);
    // Moon orbits Earth, not the Sun — use Earth's orbital elements for heliocentric position
    // (mirrors JS: key = (planetKey === 'moon') ? 'earth' : planetKey)
    let key = if planet == Planet::Moon { Planet::Earth } else { planet };
    let el = ORB_ELEMS[key.index()];

    let l = (el.l0 + el.dl * t).rem_euclid(360.0) * DEG; // mean longitude (rad)
    let w = el.om * DEG;                                    // longitude of perihelion (rad)
    let m = (l - w).rem_euclid(2.0 * PI);                  // mean anomaly (rad)
    let e = el.e0;
    let big_e = kepler_e(m, e);

    // True anomaly
    let nu = 2.0 * f64::atan2(
        ((1.0 + e) / (1.0 - e)).sqrt() * f64::sin(big_e / 2.0),
        f64::cos(big_e / 2.0),
    );

    let r   = el.a * (1.0 - e * big_e.cos());
    let lon = (nu + w).rem_euclid(2.0 * PI);
    let x   = r * lon.cos();
    let y   = r * lon.sin();
    HelioPos { x, y, r, lon }
}

// ── Distance / light travel ───────────────────────────────────────────────────

/// Euclidean distance in AU between two bodies at `utc_ms`.
pub fn body_distance_au(a: Planet, b: Planet, utc_ms: i64) -> f64 {
    let pa = helio_pos(a, utc_ms);
    let pb = helio_pos(b, utc_ms);
    let dx = pa.x - pb.x;
    let dy = pa.y - pb.y;
    (dx * dx + dy * dy).sqrt()
}

/// One-way light travel time in seconds between two bodies.
pub fn light_travel_seconds(a: Planet, b: Planet, utc_ms: i64) -> f64 {
    body_distance_au(a, b, utc_ms) * AU_SECONDS
}

// ── Line of sight ─────────────────────────────────────────────────────────────

/// Line-of-sight status between two bodies.
#[derive(Debug, Clone, Copy)]
pub struct LineOfSight {
    pub clear:           bool,
    pub blocked:         bool,   // elongation < 3°
    pub degraded:        bool,   // 3° ≤ elongation < 10°
    pub closest_sun_au:  Option<f64>,
    pub elong_deg:       f64,
}

/// Check whether the direct path between `a` and `b` passes through the Sun.
pub fn check_line_of_sight(a: Planet, b: Planet, utc_ms: i64) -> LineOfSight {
    let pa = helio_pos(a, utc_ms);
    let pb = helio_pos(b, utc_ms);

    let dx = pb.x - pa.x;
    let dy = pb.y - pa.y;
    let d2 = dx * dx + dy * dy;

    // Minimum Sun-distance along the line segment
    let closest_sun_au = if d2 < 1e-30 {
        None
    } else {
        let t = (-(pa.x * dx + pa.y * dy) / d2).clamp(0.0, 1.0);
        let cx = pa.x + t * dx;
        let cy = pa.y + t * dy;
        Some((cx * cx + cy * cy).sqrt())
    };

    // Elongation: angle at body-a between Sun and body-b
    let elong_deg = {
        let ra  = pa.r;
        let rb  = body_distance_au(a, b, utc_ms);
        let rsun = pa.r; // a–Sun = ra
        // law of cosines: cos(elong) = (ra² + rb² - rb_abs²) / 2 ra rb
        // Simpler: angle between vector (0,0)→pa and pa→pb
        let ax = -pa.x;   // Sun relative to a
        let ay = -pa.y;
        let bx = pb.x - pa.x;
        let by = pb.y - pa.y;
        let dot = ax * bx + ay * by;
        let mag = (ax * ax + ay * ay).sqrt() * (bx * bx + by * by).sqrt();
        if mag < 1e-30 { 0.0 } else { f64::acos((dot / mag).clamp(-1.0, 1.0)) / DEG }
    };
    let _ = (closest_sun_au, pa.r); // suppress unused

    let blocked  = elong_deg < 3.0;
    let degraded = !blocked && elong_deg < 10.0;
    let clear    = !blocked && !degraded;

    LineOfSight { clear, blocked, degraded, closest_sun_au, elong_deg }
}

/// Lower-quartile (p25) light travel time — 360 orbital samples over one Earth year.
pub fn lower_quartile_light_time(a: Planet, b: Planet, ref_ms: i64) -> f64 {
    let year_ms = 365_250 * 86_400i64; // approx Julian year in ms
    let step    = year_ms / 360;
    let mut samples: Vec<f64> = (0..360)
        .map(|i| light_travel_seconds(a, b, ref_ms + i as i64 * step))
        .collect();
    samples.sort_by(|x, y| x.partial_cmp(y).unwrap());
    samples[samples.len() / 4]
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn jde_at_j2000() {
        // J2000_MS = Date.UTC(2000,0,1,12,0,0) — should give JDE ≈ 2451545.0
        let j = jde(J2000_MS);
        assert!((j - 2_451_545.0).abs() < 0.01, "jde at J2000 = {}", j);
    }

    #[test]
    fn jc_at_j2000() {
        let t = jc(J2000_MS);
        assert!(t.abs() < 1e-4, "jc at J2000 = {}", t);
    }

    #[test]
    fn light_travel_earth_mars_opposition_2003() {
        // 2003-08-27 ≈ closest Mars approach; ~186 s ±15
        let utc_ms = 1_061_991_060_000i64;
        let lt = light_travel_seconds(Planet::Earth, Planet::Mars, utc_ms);
        assert!((lt - 186.0).abs() < 15.0, "E→Mars 2003-08 = {} s", lt);
    }

    #[test]
    fn light_travel_earth_mars_2020_opposition() {
        // 2020-10-13 — ~207 s ±15
        let utc_ms = 1_602_633_600_000i64;
        let lt = light_travel_seconds(Planet::Earth, Planet::Mars, utc_ms);
        assert!((lt - 207.0).abs() < 20.0, "E→Mars 2020 = {} s", lt);
    }

    #[test]
    fn light_travel_earth_jupiter() {
        // 2023-11-03 — ~2010 s ±120
        let utc_ms = 1_699_046_400_000i64;
        let lt = light_travel_seconds(Planet::Earth, Planet::Jupiter, utc_ms);
        assert!((lt - 2010.0).abs() < 120.0, "E→Jupiter = {} s", lt);
    }

    #[test]
    fn helio_r_earth_near_1au() {
        let hp = helio_pos(Planet::Earth, J2000_MS);
        assert!((hp.r - 1.0).abs() < 0.02, "Earth r = {} AU", hp.r);
    }

    #[test]
    fn helio_r_j2000_mercury_fixture() {
        // reference.json: J2000 mercury helio_r_au ≈ 0.46648
        let hp = helio_pos(Planet::Mercury, J2000_MS);
        assert!((hp.r - 0.46648).abs() < 0.001, "Mercury r = {}", hp.r);
    }

    #[test]
    fn los_earth_mars_clear() {
        // Normal date — should not be in conjunction
        let los = check_line_of_sight(Planet::Earth, Planet::Mars, J2000_MS);
        assert!(los.elong_deg > 0.0);
    }

    #[test]
    fn lower_quartile_earth_mars_positive() {
        let lq = lower_quartile_light_time(Planet::Earth, Planet::Mars, J2000_MS);
        assert!(lq > 0.0 && lq < 1300.0, "p25 = {} s", lq);
    }
}
